rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,129 @@
+
+To support these extensions, `tsh` will need to be updated to extend the current Teleport Key Agent and add a new Teleport Key Agent client. The Teleport Key Agent client can be used by `tsh` calls on a remote host to send SSH Agent requests to the Teleport Key Agent running on the local host.
+
+#### SSH Agent Extensions
+
+**Extension Structure:**
+
+Extension requests consist of:
+
+        byte            SSH_AGENTC_EXTENSION
+        string          extension type (type@domain)
+        byte[]          extension contents
+
+An Extension response can be any custom message, but failures should result in `SSH_AGENT_EXTENSION_FAILURE`. Unsupported extensions should result in an `SSH_AGENT_FAILURE` response to differentiate from actual extension failures.
+
+**TLS Certificate Extension:**
+
+This extension requests a TLS certificate from the agent matching the provided public key. Clients should provide the public key part of a Teleport certificate, since a matching TLS certificate should be available in the local agent's key store.
+
+For session requests that require per-session MFA, clients can provide the `require_mfa` flag to issue a single time use mfa-verified certificate.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          tls-cert@goteleport.com
+        string          public key SHA
+        bool            require_mfa
+
+The returned x509 certificate can be used to perform a TLS handshake.
+
+        byte            SSH_AGENT_SUCCESS
+        string          certificate contents
+
+**Trusted CAs Extension:**
+
+This extension requests a list of trusted CA certificates.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          ca-certs@goteleport.com
+
+The retured x509 certificates can be used to form a TLS CA pool.
+
+        byte            SSH_AGENT_SUCCESS
+        string[]        certificate contents list
+
+**Known Hosts Extension:**
+
+This extension requests a list of known hosts.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          known-hosts@goteleport.com
+
+The retured known hosts can be used to perform an SSH host key callback.
+
+        byte            SSH_AGENT_SUCCESS
+        string[]        known hosts contents
+
+#### SSH Agent Key Constraint Extensions
+
+For Per session MFA, we *could* add an `ssh-cert` extension with the same structure as the `tls-cert` extension above, but we'd be missing a great opportunity. Instead, we can add a new [key constraint extension](https://datatracker.ietf.org/doc/html/draft-miller-ssh-agent#section-4.2.6.3) to add per-session MFA functionality into the Teleport Key Agent in a way that would work with `ssh` as well. This would extend per-session MFA functionality to OpenSSH integrations, including [OpenSSH Proxy Jump](https://github.com/gravitational/teleport/issues/17190).
```

Comment:
This is really smart

Reply:
Thanks! Unfortunately I had to add some caveats though.